tests/utilities/test_auto_restart.py

```diff
@@ -0,0 +1,594 @@
+    """
+    This test ensures ``FastForwardSampler`` and ``CaptureIterableDataset`` are properly being
+    used to capture workers states.
+    """
+    batch_size = 3
+    initial_seed = seed_everything(42)
+    generator = torch.Generator()
+    generator.manual_seed(initial_seed)
+    dataset = RangeIterableDataset(range(20), num_workers, batch_size, True)
+    dataset = CaptureIterableDataset(dataset, num_workers)
+    dataloader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, generator=generator)
+    iter_dataloader = iter(dataloader)
+    batches = []
+    for _ in range(5):
+        batches.append(next(iter_dataloader))
+
+    if num_workers == 0:
+        batch_0_expected = torch.tensor([4, 3, 12])
+        batch_1_expected = torch.tensor([18, 0, 7])
+        batch_2_expected = torch.tensor([8, 14, 11])
+        batch_3_expected = torch.tensor([5, 10, 13])
+        batch_4_expected = torch.tensor([17, 19, 15])
+    elif num_workers == 1:
+        batch_0_expected = torch.tensor([3, 18, 17])
+        batch_1_expected = torch.tensor([13, 2, 19])
+        batch_2_expected = torch.tensor([6, 4, 7])
+        batch_3_expected = torch.tensor([1, 14, 5])
+        batch_4_expected = torch.tensor([12, 8, 16])
+    else:
+        batch_0_expected = torch.tensor([3, 4, 5])
+        batch_1_expected = torch.tensor([10, 12, 14])
+        batch_2_expected = torch.tensor([7, 0, 9])
+        batch_3_expected = torch.tensor([16, 18, 17])
+        batch_4_expected = torch.tensor([8, 1, 2])
+
+    assert torch.equal(batches[0]["data"], batch_0_expected)
+    assert torch.equal(batches[1]["data"], batch_1_expected)
+    assert torch.equal(batches[2]["data"], batch_2_expected)
+    assert torch.equal(batches[3]["data"], batch_3_expected)
+    assert torch.equal(batches[4]["data"], batch_4_expected)
+
+    # restarting on batch_1 and getting 3 extra batches
+
+    state_dict = {'iter_sampler': {}}
+    for batch in batches[:2]:
+        _state_dict = CaptureIterableDataset.convert_batch_into_state_dict(batch)
+        for k, v in _state_dict.items():
+            state_dict[k].update(v)
+
+    assert len(state_dict["iter_sampler"]) == (num_workers if num_workers > 1 else 1)
+
+    initial_seed = seed_everything(42)
+    generator.manual_seed(initial_seed)
+    dataset = RangeIterableDataset(range(20), num_workers, batch_size, True, state_dict=state_dict)
+    dataset = CaptureIterableDataset(dataset)
+    dataset.load_state_dict(state_dict)
+    dataloader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, generator=generator)
+    iter_dataloader = iter(dataloader)
+    batches = []
+    for _ in range(3):
+        batches.append(next(iter_dataloader))
+
+    assert torch.equal(batches[0]["data"], batch_2_expected)
+    assert torch.equal(batches[1]["data"], batch_3_expected)
+    assert torch.equal(batches[2]["data"], batch_4_expected)
+
+
+def _setup_ddp(rank, worldsize):
```

Comment:
Do you need to destroy the process group after init?

Reply:
I think it should be killed automatically as it is using spawned right ?